src/main/java/org/jitsi/jibri/service/impl/FileRecordingJibriService.kt

```diff
@@ -0,0 +1,158 @@
+import org.jitsi.jibri.util.ProcessMonitor
+import org.jitsi.jibri.util.WriteableDirectory
+import org.jitsi.jibri.util.extensions.error
+import java.io.File
+import java.io.IOException
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import java.util.logging.Logger
+
+data class RecordingOptions(
+    /**
+     * The directory in which recordings should be created
+     */
+    val recordingDirectory: WriteableDirectory,
+    /**
+     * The params needed to join the call
+     */
+    val callParams: CallParams,
+    /**
+     * The filesystem path to the script which should be executed when
+     *  the recording is finished.
+     */
+    val finalizeScriptPath: String
+)
+
+/**
+ * Set of metadata we'll put alongside the recording file(s)
+ */
+data class RecordingMetadata(
+    val participants: List<String>
+)
+
+/**
+ * [FileRecordingJibriService] is the [JibriService] responsible for joining
+ * a web call, capturing its audio and video, and writing that audio and video
+ * to a file to be replayed later.
+ */
+class FileRecordingJibriService(private val recordingOptions: RecordingOptions) : JibriService() {
+    /**
+     * The [Logger] for this class
+     */
+    private val logger = Logger.getLogger(this::class.qualifiedName)
+    /**
+     * Used for the selenium interaction
+     */
+    private val jibriSelenium = JibriSelenium(
+        JibriSeleniumOptions(recordingOptions.callParams),
+        Executors.newSingleThreadScheduledExecutor(NameableThreadFactory("JibriSelenium"))
+    )
+    /**
+     * The [FfmpegCapturer] that will be used to capture media from the call and write it to a file
+     */
+    private val capturer = FfmpegCapturer()
+    /**
+     * The [Sink] this class will use to model the file on the filesystem
+     */
+    private var sink: Sink
+    /**
+     * If ffmpeg dies for some reason, we want to restart it.  This [ScheduledExecutorService]
+     * will run the process monitor in a separate thread so it can check that it's running on its own
+     */
+    private val executor: ScheduledExecutorService =
+        Executors.newSingleThreadScheduledExecutor(NameableThreadFactory("FileRecordingJibriService"))
+    /**
+     * The handle to the scheduled process monitor task, which we use to
+     * cancel the task
+     */
+    private var processMonitorTask: ScheduledFuture<*>? = null
+
+    init {
+        sink = FileSink(recordingOptions.recordingDirectory, recordingOptions.callParams.callUrlInfo.callName)
+        jibriSelenium.addStatusHandler {
+            publishStatus(it)
+        }
+    }
+
+    /**
+     * @see [JibriService.start]
+     */
+    override fun start(): Boolean {
+        if (!jibriSelenium.joinCall(recordingOptions.callParams.callUrlInfo.callName)) {
+            logger.error("Selenium failed to join the call")
+            stop()
+            return false
```

Comment:
Without much context, this seems like an "exceptional" case. When do you decide, in general, to throw an error versus return a status?